Generate only kpis

marota opened the pull request:
You can now generate KPIs based on existing data as with this cli
 ```
 chronix2grid --mode K --output-folder pathTo/data_grid2op/l2rpn_wcci_2022/chronics --input-folder ChroniX2Grid_path/input_data --ignore-warnings --weeks 4 --case case118_l2rpn_wcci_2022 --n_scenarios 1 
```
You can also only generate KPIs for your generated data if you don't want to bother generating KPIs for some reference data. This can be done by setting in ParamsKPI.json
 ```
"comparison":null 
 ```

BDonnot commented on input_data/generation/case118_l2rpn_wcci_2022/remove_load.py:
```diff
@@ -0,0 +1,24 @@
+import pandapower
+import pandapower as pp
+import numpy as np
+import pdb
+grid = pp.from_json("grid.json")
+idx_of_element_to_delete = np.isin(grid.load.name,
+                                   ["load_7_6",
+                                    "load_23_20",
+                                    "load_71_56",
+                                    "load_72_57",
+                                    "load_90_72",
+                                    "load_98_80",
+                                    "load_112_93",
+                                    "load_115_96"])
+idx_of_element_to_delete = np.where(idx_of_element_to_delete)[0]
+grid.load.drop(idx_of_element_to_delete, inplace=True)
+grid.load.reset_index(inplace=True, drop=True)
+pp.runpp(grid)
+p_mw = 0.
+max_e_mwh = 24.
+for bus in [22, 41, 44, 58, 76, 95, 112]:
+    pp.create_storage(grid, bus, p_mw, max_e_mwh)
+pp.runpp(grid)
+pp.to_json(grid, "grid.json")
```
probably not great to have this file here either

marota replied:
Removed

BDonnot commented on requirements.txt:
```diff
@@ -8,6 +8,7 @@ ipywidgets>=7.5.1
 matplotlib==3.1.3
+networkx==2.8.8
 numba>=0.48.0
 numpy==1.18.3
```
completely outdated (I modified the code in the branch `ramp_forecast` to use up to date versions of these libraries, I can make the PR and we merge this after if you prefer.

marota replied:
removed

BDonnot commented on tests/data/input/generation/case118_l2rpn_wcci_2022/gen_info.txt:
```diff
@@ -0,0 +1,94 @@
+eolien en mer: pmax = 250
+solaire: 93 ou 80
+eolien terrestre : 84.0
+
+
+### Adapter la prod a rapport RTE 2050
+Par rapport au réseau d'avant:
+- enlevé les "petits" thermiques (pmax 50. ou 100.), remplacés par du solaire
+- augmenté le nombre de "gros" thermiques (gaz avec CSC)
+- mis du "gros éolien" à gauche (pour "simuler" des cotes)
+- gen sub 11 => gros éolien (était hydro), gen sub 14 => hydro (était thermique)
```
this file has nothing to do here either

marota replied:
Removed

BDonnot reviewed (changes requested):
I think some changes need to be made before merging, most notably:

- the plt.ioff() or something like that, which can have side effects impossible to find and debug
- the requirements.txt that uses completely outdated library (in the branch `ramp_forecasts` i made the changes to use most recent versions and tests pass, maybe we could merge this branch first ?)
- some minor readability issues with the `if not`

marota commented on README.md:
````diff
@@ -182,6 +188,10 @@ Some general parameters have to be set in *INPUT_FOLDER/kpi/paramsKPI.json*
 
+**Note**: if you just want to generate the KPIs from your synthetic data without comparison to reference chronics, you can set 
+```json
+"comprison":null
````
Ok done

marota commented on README.md:
````diff
@@ -82,7 +82,13 @@ To provide --input-data and --output-folder folders and specify --case grid case
 ```
-
+#### KPIs generation with existing chronics as in grid2op env
+If you already generated your data such as from a grid2op environment, you can generate KPIs afterwards under this cli given that a *paramsKPI.json* exists in the *input-folder/kpi/grid_case*
+ ```commandline
+ chronix2grid --mode K --output-folder pathTo/data_grid2op/l2rpn_wcci_2022/chronics --input-folder ChroniX2Grid_path/input_data --ignore-warnings --weeks 4 --case case118_l2rpn_wcci_2022 --n_scenarios 1 
+```
+In your "chronics" folder, this will create a "kpi" folder with subfolders of KPIs per chronic 
+ 
````
I created an issue for improvement to be able to directly do this from grid2op indeed

marota commented on chronix2grid/config.py:
```diff
@@ -166,7 +166,33 @@ def read_configuration(self):
+                sep=',')
+            names = loads_charac['name']  # to generate error if separator is wrong
+
+        except:
```
Good point, it was already there before. As I am not sure which exception to use here and it is not per say related to this PR, I create this issue to keep track of it https://github.com/BDonnot/ChroniX2Grid/issues/81

marota commented on chronix2grid/kpi/deterministic/kpis.py:
```diff
@@ -21,6 +21,7 @@
 from plotly.subplots import make_subplots
 import seaborn as sns
 
+plt.ioff() #to not launch windows with plot in os
```
Ok not sure why it was there, I commented it

marota commented on chronix2grid/kpi/deterministic/kpis.py:
```diff
@@ -87,9 +90,15 @@ def __init__(self, ref_consumption, syn_consumption, ref_dispatch,
 
         # Aggregate chronics variables
-        self.ref_agg_conso = ref_consumption.sum(axis=1)
+        if not self.kpi_on_syn_data_only:
+            self.ref_agg_conso = ref_consumption.sum(axis=1)
+        else:
+            self.ref_agg_conso=None
```
The most general case comes first as you would expect normally some reference data for comparison. I would keep it like that for this reason, even if I agree with you in genera

marota commented on chronix2grid/main.py:
```diff
@@ -205,19 +226,18 @@ def generate_inner(case, start_date, weeks, by_n_weeks, n_scenarios, mode,
-    elif 'T' in mode and 'K' in mode:
-        # Get and format dispatched chronics, then compute KPI and save plots
-        wind_solar_only = False
+        print("WARNING: make sure that your generated data is in folder: "+generation_output_folder)
```
Good point, I added an issue https://github.com/BDonnot/ChroniX2Grid/issues/82

marota commented on chronix2grid/kpi/deterministic/kpis.py:
```diff
@@ -143,20 +154,31 @@ def _plot_heatmap(self, corr, title, path_png=None, save_png=True):
-        if every_nth is not None:
-            for i in range(2):
-                for n, label in enumerate(axes[i].xaxis.get_ticklabels()):
+        plt.ioff()  # to not launch windows with plot in os
```
I commented it as not sure what it is used for

marota commented on chronix2grid/kpi/main.py:
```diff
@@ -56,67 +60,109 @@ def main(kpi_input_folder, generation_output_folder, scenario_names,
-                scenario_generation_output_folder, kpi_input_folder, year,
-                prods_charac, loads_charac, wind_solar_only,
-                params, case)
+        #if wind_solar_only:
+        #    # Get reference and synthetic dispatch and loads
+        #    (ref_dispatch, ref_consumption, syn_dispatch, syn_consumption,
+        #     paramsKPI) = pivot_format(
+        #        scenario_generation_output_folder, kpi_input_folder, year,
+        #        prods_charac, loads_charac,
+        #        params, case,wind_solar_only,kpi_on_syn_data_only)
+        #    ref_prices = None
+        #    prices = None
+        #
```
Ok removed it

marota commented on chronix2grid/kpi/main.py:
```diff
@@ -56,67 +60,109 @@ def main(kpi_input_folder, generation_output_folder, scenario_names,
+                                                                                          kpi_input_folder, year,
+                                                                                          prods_charac, loads_charac, params, case,
+                                                                                          wind_solar_only)
+        #else:
+        #    # Get reference and synthetic dispatch and loads
+        #    (ref_dispatch, ref_consumption, syn_dispatch, syn_consumption,
+        #     ref_prices, prices, paramsKPI) = pivot_format(
+        #        scenario_generation_output_folder, kpi_input_folder, year,
+        #        prods_charac, loads_charac,
+        #        params, case,wind_solar_only,kpi_on_syn_data_only)
```
Ok removed it

marota commented on chronix2grid/kpi/main.py:
```diff
@@ -165,3 +211,176 @@ def main(kpi_input_folder, generation_output_folder, scenario_names,
+                "format."
+            )
+        time_interval = timedelta(hours=tmp.hour, minutes=tmp.minute).total_seconds()/60
+    return start_datetime,time_interval
```
Definitely beeter thanks and done

marota commented on chronix2grid/kpi/preprocessing/pivot_KPI.py:
```diff
@@ -43,25 +61,20 @@ def pivot_format(chronics_folder, kpi_input_folder, year, prods_charac, loads_ch
-    timestep = paramsKPI['timestep']
+    kpi_case_input_folder = None
+
+    if kpi_input_folder and os.path.exists(os.path.join(kpi_input_folder, case)):
+        kpi_case_input_folder= os.path.join(kpi_input_folder, case)
```
ok merci, done

marota commented on chronix2grid/kpi/preprocessing/pivot_KPI.py:
```diff
@@ -88,15 +101,58 @@ def pivot_format(chronics_folder, kpi_input_folder, year, prods_charac, loads_ch
+    prices=None
     if wind_solar_only:
         syn_prod, syn_load = chronics_to_kpi(chronics_folder, timestep, params,
                                              thermal=not wind_solar_only)
-        return ref_prod, ref_load, syn_prod, syn_load, paramsKPI
+        return syn_prod, syn_load,prices
     else:
         syn_prod, syn_load, prices = chronics_to_kpi(
             chronics_folder, timestep, params, thermal=not wind_solar_only)
-        return ref_prod, ref_load, syn_prod, syn_load, ref_prices, prices, paramsKPI
+        return syn_prod, syn_load, prices
```
Yes indeed there was some unecessary code here, i changed it with your proposal

marota commented on chronix2grid/kpi/preprocessing/pivot_utils.py:
```diff
@@ -363,8 +362,17 @@ def chronics_to_kpi(chronics_repo, timestep, params, thermal = True):
 
         load_p = pd.read_csv(os.path.join(chronics_repo, 'load_p.csv.bz2'), sep=';', decimal='.')
```
Good point, done

marota commented on chronix2grid/kpi/preprocessing/pivot_utils.py:
```diff
@@ -350,10 +347,12 @@ def chronics_to_kpi(chronics_repo, timestep, params, thermal = True):
-        price['Time'] = datetime_index[:len(price)]
+        path_file_prices=os.path.join(chronics_repo, 'prices.csv.bz2')
+        if(os.path.exists(path_file_prices)):
+            price = pd.read_csv(os.path.join(chronics_repo, 'prices.csv.bz2'),
```
Indeed, done

marota commented on chronix2grid/main.py:
```diff
@@ -99,7 +111,8 @@ def generate_mp_core(prng, case, start_date, weeks, by_n_weeks, n_scenarios, mod
     print('initial_seeds')
     print(initial_seeds)
-    dump_seeds(generation_output_folder, initial_seeds, scenario_name)
+    if mode!='K':#if data is generated
```
Yes it is really see, meaning if we are in a mode in which data is generated and not only computing kpis
I extended a bit the comment for better clarity there

marota commented on chronix2grid/main.py:
```diff
@@ -111,26 +124,31 @@ def generate_mp_core(prng, case, start_date, weeks, by_n_weeks, n_scenarios, mod
-    rm_temporary_folders(input_folder, case)
-
-def rm_temporary_folders(input_folder, case):
+        pool = multiprocessing.Pool(nb_core)
+        iterable = [i for i in range(n_scenarios)]
+        multiprocessing_func = partial(
+            generate_per_scenario,
+            case, start_date, weeks, by_n_weeks, mode, input_folder,
+            kpi_output_folder, generation_output_folder, scen_names,
+            seeds_for_loads, seeds_for_res, seeds_for_disp, ignore_warnings)
+
+        pool.map(multiprocessing_func, iterable)
+        pool.close()
```
Ok thanks done !

marota commented on tests/data/input/generation/case118_l2rpn_wcci_2022/remove_load.py:
```diff
@@ -0,0 +1,24 @@
+import pandapower
```
Removed

marota commented on input_data/generation/case118_l2rpn_wcci_2022/gen_info.txt:
```diff
@@ -0,0 +1,94 @@
+- augmenté le nombre de "gros" thermiques (gaz avec CSC)
+- mis du "gros éolien" à gauche (pour "simuler" des cotes)
+- gen sub 11 => gros éolien (était hydro), gen sub 14 => hydro (était thermique)
+- remplacer les thermiques "moyens" (rampes 4.3 et 5.4) et par du nucléaire (ça serait des SMR)
```
Removed

marota commented:
@BDonnot thanks for the long review, I addressed all comments and merged master into it to be up to date.
I think this PR can be merged, so you can have a last quick check
